xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,570 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.Listener;
+import io.envoyproxy.envoy.api.v2.RouteConfiguration;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.api.v2.route.Route;
+import io.envoyproxy.envoy.api.v2.route.VirtualHost;
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.Bootstrapper.ChannelCreds;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class XdsClientImpl extends XdsClient {
+  private static final Logger logger = Logger.getLogger(XdsClientImpl.class.getName());
```

Comment:
Consider using ChannelLogger instead. It's associated with channe's InternalLogId, and anything above DEBUG level also goes to channelz.

Reply:
`ChannelLogger` is only accessible through `LoadBalancer.Helper`. We can't have a `ChannelLogger` as one of `XdsClientImpl`'s constructor argument, the classes instantiating `XdsClientImpl` may not be able to get `ChannelLogger`.

For the xDS channel, there is no way to get its `ChannelLogger`.

Maybe we should consider some API changes on `ManagedChannel` to make `ChannelLogger` more accessible?

Reply:
Where would `XdsClientImpl` be instantiated?

Reply:
The xDS resolver is one place for sure. It creates an `XdsClient` instance and start doing LDS/RDS to received `balancerName` and craft a service config, which will be passed to LB policies. This `XdsClient` instance will also be passed to LB policies so that LB policies are able to put CDS/EDS watchers to receive information from CDS/EDS responses. You can see this usage in [C++'s implementation here](https://github.com/grpc/grpc/blob/6dde2f43f70e10759bff06e09207d48f66beff36/src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc#L90-L92).

According to @ejona86, LB policies should also be able to create `XdsClient` for their own use when they do not find the reference of `XdsClient` in `ResolvedAddresses.Attributes`. You can see this usage in [C++'s implementation here](https://github.com/grpc/grpc/blob/6dde2f43f70e10759bff06e09207d48f66beff36/src/core/ext/filters/client_channel/lb_policy/xds/xds.cc#L794-L798).

`XdsClient` encapsulate all the logic for LDS/RDS/CDS/EDS. The xDS resolver only interested in results of LDS/RDS, the CDS LB policy only interested in results of CDS responses, and the EDS LB policy only interested in results of EDS responses. I believe the design is that, we do not require all of xDS resolver, CDS LB policy and EDS LB policy to present to make things work. For example, if we only implemented EDS LB policy, we should still be able to do EDS only with `XdsClient`.

Reply:
We can keep discussion and for this PR let's make a TODO item (as well as a github issue) for the `ChannelLogger` issue for now.

Reply:
Regarding LB policies instantiating `XdsClient` do they have all the values to create it (such as `serverUri`, `node`,     `ChannelCreds`)?

Reply:
These parameters come from `BootstrapInfo` after reading the bootstrap file.